src/widgets/RealtimeEffectPanel.cpp

```diff
@@ -0,0 +1,711 @@
+
+         //Show effect settings
+         auto optionsButton = safenew ThemedButtonWrapper<wxButton>(this, wxID_ANY, wxT(""), wxDefaultPosition, wxDefaultSize, wxNO_BORDER);
+         optionsButton->SetBitmapLabelIndex(bmpEffectSettingsNormal);
+         optionsButton->SetBitmapPressedIndex(bmpEffectSettingsDown);
+         optionsButton->SetBitmapCurrentIndex(bmpEffectSettingsHover);
+         optionsButton->SetBackgroundColorIndex(clrEffectListItemBackground);
+         optionsButton->SetMinSize(optionsButton->GetBitmapLabel().GetSize());
+         optionsButton->Bind(wxEVT_BUTTON, &RealtimeEffectControl::OnOptionsClicked, this);
+         
+         auto dragArea = safenew wxStaticBitmap(this, wxID_ANY, theTheme.Bitmap(bmpDragArea));
+         dragArea->Disable();
+         sizer->Add(
+            dragArea,
+            wxSizerFlags().Center().Border(wxLEFT, 5)
+         );
+         sizer->Add(
+            mEnableButton,
+            wxSizerFlags().Center().Border(wxLEFT, 5)
+         );
+         sizer->Add(
+            mChangeButton,
+            wxSizerFlags().Center().Border(wxLEFT, 5).Proportion(1)
+         );
+         sizer->Add(
+            optionsButton,
+            wxSizerFlags().Center().Border(wxLEFT | wxRIGHT, 5)
+         );
+
+         SetSizer(sizer.release());
+      }
+
+      void SetEffect(const std::shared_ptr<Track>& track, int effectIndex)
+      {
+         mTrack = track;
+
+         auto& effects = RealtimeEffectList::Get(*track);
+         assert (effectIndex >= 0 && effectIndex < effects.GetStatesCount());
+
+         mEffectState = effects.GetStateAt(effectIndex);
+         auto desc = PluginManager::Get().GetPlugin(mEffectState->GetID());
+
+         mChangeButton->SetLabel(desc->GetSymbol().Translation());
+      }
+
+      void RemoveFromList()
+      {
+         if(mEffectState != nullptr)
+         {
+            auto& effects = RealtimeEffectList::Get(*mTrack);
+            effects.RemoveState(*mEffectState);
+         }
+
+         wxPostEvent(GetParent(), wxCommandEvent (EVT_EFFECT_ITEM_REMOVE));
+         
+         Destroy();
+      }
+      
+      void OnOptionsClicked(wxCommandEvent& event)
+      {
+         if(mEffectState == nullptr)
+            return;//not initialized
+         
+         auto effectPlugin = EffectManager::Get().GetEffect(mEffectState->GetID());
+         
+         auto access = mEffectState->GetAccess();
+            
+         access->ModifySettings([&](EffectSettings& settings)
+         {
+            auto project = mTrack->GetOwner()->GetOwner();
+            auto &trackFactory = WaveTrackFactory::Get( *project );
+            auto &selectedRegion = ViewInfo::Get( *project ).selectedRegion;
+            auto rate = ProjectRate::Get(*project).GetRate();
+            auto &window = ProjectWindow::Get( *project );
+            auto flags = static_cast<unsigned>(EffectManager::kNone);
+
+            effectPlugin->DoEffect(settings,
```

Comment:
Still more Undo stack pushes are needed, for changes of settings in an effect stack.

But this doesn't work -- as you may have known from testing.  I think it was something you needed to discuss with me, but didn't find the opportunity?

It doesn't work because for realtime capable effects, a non-modal dialog is opened, and ModifySettings exits before the dialog closes.

Maybe then, the individual effect has other ModifySettings calls in its event handlers, so it sort-of works.  But not as you might expect from reading this code, and it won't interact correctly with undo and redo.

How to fix this then?  See the example of `EffectManager::PromptUser` and `MacroCommands::PromptForParamsFor` which calls it, while inside this scope object's lifetime:

```
   auto cleanup = EffectManager::Get().SetBatchProcessing(ID);
```

That forces effect dialogs to open modally, even for realtime effects, if you are editing a macro, which is how you can step through those functions and understand.

But they change the global EffectSettings for the effect, which is of course not the right thing here.

Reply:
Another question to ask is whether all these things requiring pushes and pops of the undo stack are safe to do while playback is in progress.

I think we surely want that -- one hope is that you listen as you change settings of a single effect, or even change the composition of the effect list.  But now I doubt whether all is set up correctly in the audio engine for that, both to avoid dangling pointers and crashes, and to communicate changes of the sequence of effects properly inter-thread.

The limited thing that I know was anticipated by Leland's work, is the powering on and off of individual effects, which is what the suspending and resuming of states is for.

And the point of the EffectSettingsAccess class I developed is to communicate changes of settings of an individual effect correctly -- what this piece of code here will do.

But adding, deleting, and permuting effects?  Or changing the whole stack at one with undo and redo?

I must think.